src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,162 @@
+            next.prev = newNode;
+            head = newNode;
+            newNode.next = next;
+        } else {
+            next.prev.next = newNode;
+            newNode.prev = next.prev;
+            newNode.next = next;
+            next.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null) {
+            for (T node : list) {
+                add(node);
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        Node<T> node = getNode(index);
+        return node.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndex(index);
+        Node<T> nodeByIndex = getNode(index);
+        T oldValue = nodeByIndex.item;
+        nodeByIndex.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        Node<T> node = getNode(index);
+        return unlink(node);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        int index = 0;
+        while (node != null) {
+            if (node.item == object || node.item != null && node.item.equals(object)) {
+                unlink(getNode(index));
+                return true;
+            }
+            node = node.next;
+            index++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void validateIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index " + index + " for size " + size);
+        }
+    }
+
+    private void addLast(Node<T> newNode) {
+        newNode.prev = tail;
+        tail.next = newNode;
+        tail = newNode;
+        size++;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> node;
+        if (index < size / 2) {
+            node = head;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+        } else {
+            node = tail;
+            for (int i = size; i > (index + 1); i--) {
+                node = node.prev;
+            }
+        }
+        return node;
+    }
+
+    private T unlink(Node<T> node) {
+        if (node == head) {
+            if (size != 1) {
+                head.next.prev = null;
+                head = head.next;
+            } else {
+                head = null;
+                tail = null;
+            }
+        } else if (node == tail) {
+            tail.prev.next = null;
+            tail = tail.prev;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        size--;
```

Comment:
move `size--` in the start of method, and call `!isEmpty()` instead of `size != 1`